Random stuff to improve.

sellout opened the pull request.

sellout committed:
Minor clarification.

sellout committed:
Random stuff to improve.

FintanH commented on Outline.org:
```diff
@@ -0,0 +1,89 @@
+opposite of what "bottom" and "up" mean in nature. So we have to decide on a
+consistent directional metaphor (which is sorely lacking in the field as it is).
+
+* Advanced Topics
```
Other topics:
* Relation to category theory

Use cases that we could explore:
* Schemas
* Directory
* Expression language

FintanH commented on Outline.org:
```diff
@@ -0,0 +1,89 @@
+
+** top-down data passing
+
+This is the reason to not introduce unfolding earlier -- people often think of
```
Something interesting to note here: people are actually used to this style by passing an "accumulator". Another way this sounds familiar is difference lists.

sellout committed:
Updates to recent additions.

sellout committed:
Collecting some stuff that got left behind

sellout committed:
Add third part to the book.

It's now

1. introduction to concepts
2. refinement and practical application
3. advanced notions

This adds the second one, and notes for a section that talks about integrating
recursion schemes with existing code.

FintanH commented on Graphs/README.md:
```diff
@@ -0,0 +1,188 @@
+
+**TODO**: Need to split this (and other sections) into "using recursion schemes" and "understanding recursion schemes" somehow.
+
+So, let's talk about how we want to deal with graphs ... first, we'd like to maintain as much as possible the same abstractions we use when dealing with trees. That is, algebras and pattern functors. This means that given the same pattern functor, we could build either a tree or a graph (just as we can build either a finite or infinite structure from the same pattern functor). And we'd like to be ...
```
I think we could take inspiration from 7 Sketches using the same idea of some problem statement. So in this case, we should pick some problem (or set of problems) that can be solved using graphs. This will provide motivation as to why we introduce graphs. Maybe since we're comparing them to trees we can show how trees might not suffice for this problem, or how it can be generalised from a tree to a graph, etc.

I also think that more introduction definitions graphs, adjacency lists, and anything else would be helpful. Lot's of diagrams will be cool here because graphs are very visual creatures :smile:

sellout replied:
The "standard" example for graphs here is trees with bindings (like progarmming languages). On the one hand, programmers should be familiar with that notion, on the other, the problems to solve with them may be too compiler-y.

Perhaps the way we used them at Formation for offers would be a better example?

FintanH replied:
> Perhaps the way we used them at Formation for offers would be a better example?

Probably best to avoid that in case of IP conflict or something :sweat_smile: What about doing calculations on networks where peers are connected? (Not sure what inspired this :stuck_out_tongue_closed_eyes:)

sellout replied:
Well, I wouldn't want it to _be_ the Formation stuff, but just something like how we encoded the logic.

I'd be interested in the network stuff, but it's important (at least for early examples) that we're talking about finite directed acyclic graphs, and I imagine peer-to-peer fails at least one or two of those (but maybe not for specific use cases).

FintanH replied:
> Well, I wouldn't want it to be the Formation stuff, but just something like how we encoded the logic.

Ya, cool just making sure :smile: Sandy had a scavenger hunt example in his ADD book that was similar in fashion. If we wanted inspiration from him.

> we're talking about finite directed acyclic graphs

Ah right. We'd get the finite part by looking at a portion of the graph, say with a cutoff. Acyclic would not be guaranteed. But we _could_ look at something like Pijul for inspiration since changesets are DAGs :) I'm sure there are lots of examples we could list out and pick one or two :grin:

FintanH commented on Graphs/README.md:
```diff
@@ -0,0 +1,188 @@
+
+-- | Used internally to hold information we need while constructing a graph from
+--   trees.
+data GraphState f
```
I think we could spin more story around `GraphState`, eliminateCommonSubtree`, and `findOrAddNode`. So again, what's the motivation for each step. What are the smaller sub-problems we're trying to solve? <--- this is one of our general motifs :smile:

FintanH commented on Graphs/README.md:
````diff
@@ -0,0 +1,188 @@
+      pure index
+```
+
+## Benefits
````
Ah, so some of the comparisons to trees are here! This kind of motivation should come closer to the start of a section like this. Or at least _tease_ it.

FintanH commented on Graphs/README.md:
```diff
@@ -0,0 +1,188 @@
+
+The problem is that we're not treating the graph like a graph. Let's look at the type of that algebra `f (m a) -> m a`. Earlier we discussed how each node in a graph is only processed once, no matter how many times it's referenced. But, in this case, what is the result of processing that node? It's an `m a`.
+
+Basically, even though we only generate a single term of type m a for each node, that node may have it's monadic effect evaluated multiple times (once for each node it's a child of). But, we should expect each node in a graph to have its effect evaluated exactly once. Imagine if the monad is some `State` that attaches an identifier to each node in the graph. When you evaluate that `State` twice,  ...
```
```suggestion
Basically, even though we only generate a single term of type `m a` for each node, that node may have its monadic effect evaluated multiple times (once for each node it's a child of). But, we should expect each node in a graph to have its effect evaluated exactly once. Imagine if the monad is some `State` that attaches an identifier to each node in the graph. When you evaluate that `State` twice, you end up attaching two different identifiers to the same node, depending on where it's referenced from, and that's just not going to work. You might think "well, I can have the `State` track whether we've already seen this node and, if so, return the same identifier," but
```

FintanH reviewed (commented):
Can't say that I fully followed the explanations about the problems involving algebras over their graphs and their nodes :sweat_smile: 

Left some notes on how to make that section flow better.

sellout committed:
Merge pull request #4 from sellout/more-stuff

More stuff
